doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+
+```yml
+# traits.yml
+Camera:
+  description: >
+    The entity represents a device that captures a view through
+    re-projection or encoding into some other form.
+  usage:
+    - entity
+  version: 3
+```
+
+```python
+# Manager
+# In this example, CameraTrait is on version 3
+# SpatialTrait is on version 2
+#
+# The manager is responding to a resolve, needing to populate a
+# TraitsData with a camera trait and a spatial trait of the expected
+# version
+cameraVersion = mediacreation.versionFromTraitId(mediacreation.CameraTrait.kId)
+spatialVersion = mediacreation.versionFromTraitId(mediacreation.SpatialTrait.kId)
+
+# Support all known major versions.
+# Camera
+success_result = TraitsData()
+if (cameraVersion >= 3 and cameraVersion < 4):
+     trait = mediacreation.CameraTrait(success_result)
+     ... # Up to date camera population
+elif (cameraVersion >= 2 and cameraVersion < 1):
+     trait = mediacreation.legacy.CameraTrait_v2(success_result)
+     ... # V2 camera population
+elif (cameraVersion >= 1):
+     trait = mediacreation.legacy.CameraTrait_v1(success_result)
+     ... # V1 camera population
+else:
+  raise RuntimeError()
+
+# Spatial
+if (spatialVersion >= 2 and spatialVersion < 3):
+    trait = mediacreation.SpatialTrait(success_result)
+    ... # Up to date spatial population
+elif (spatialVersion >= 1):
+    trait = mediacreation.legacy.SpatialTrait_v1(success_result)
+     ... # V1 to date spatial population
+else:
+  raise RuntimeError()
+```
+
+### Benefits
+- Maximally flexible, hosts and managers can disregard versioning for
+  traits they don't care about.
+- Obvious from the YAML which version a trait is.
+- Individual traits will only update version when they change, thus they
+  will not have "gaps", they will always go v1->v2->v3.
+- A single trait can be updated in isolation without any effect on other
+  traits in the package.
+- Semver could be used to provide additional runtime compatibility
+  information.
+
+### Drawbacks
+- Version has to be encoded in `traitID` because several methods
+  (including `resolve`) take `traitSet`, which is only a collection of
+  `traitIDs`. There is no `TraitsData` to inspect. This would be a major
+  breaking change as currently conceived.
+- Version support may become fragmented if hosts and managers provide
+  versioned support for only certain traits
+- When packing multiple traits, manager has to branch on each trait as
+  they could be different versions.
+- When extended to specification this _really_ causes problems, as we
+  need to invent a new place to pack specifications version across the
+  wire.
+
+## Version the whole schema.
+This option places a version number is placed at the top of the
+`traits.yml` file or equivalent. This defines the version for all the
+traits within the package.
+
+The manager will still need a mechanism to know which trait-library
+version the host will be expecting. This could be achieved by a
+`HostInterface` method.
```

Comment:
We did discover that the `HostInterface` is available at `initialize()` time, so the manager _could_ configure its adaptation as part of its initialization, avoiding the need to make the call with each request (and sidestep the out-of-process callback issue).

Another solution would be to add a new argument to `initialize` that advertises the trait schema versions in use by the host.

If we require trait schema version negotiation between host and manager (see ), then I suspect we'd need a different set of functions altogether (or perhaps the `initialize` process could incorporate this). Also see

Reply:
The concern is more about making requests more stateful. Which in many scenarios then forces a custom manager state object.

Reply:
As an example, https://github.com/OpenAssetIO/OpenAssetIO/issues/1203 feels it would get a lot more complex/challenging with out-of-band version coordination.